### ISSUE TITLE
Synapse returns a 500 error when requesting /state or /state_ids over federation with no event_id

### ISSUE THREAD

**Issue by turt2live** [matrix-org/synapse#3646]
eg: `GET /_matrix/federation/v1/state/!room:domain.com/` (lack of `?event_id=$whatever`)

This should probably be a 400 error, which appears to be the intention given the stacktrace. However, a 500 is returned.

```
synapse.http.server - 101 - ERROR - GET-74- Failed handle request via <function _async_render at 0x7f0817ae3e60>: <XForwardedForRequest at 0x7f0729263cf8 method='GET' uri='/_matrix/federation/v1/state/!redacted/' clientproto='HTTP/1.0' site=8448>: Traceback (most recent call last):
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/federation/transport/server.py", line 227, in new_func
    origin, content, request.args, *args, **kwargs
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/federation/transport/server.py", line 343, in on_GET
    query.get("event_id", [None])[0],
  File "/root/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/root/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
--- <exception caught here> ---
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/http/server.py", line 78, in wrapped_request_handler
    yield h(self, request)
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/http/server.py", line 304, in _async_render
    callback_return = yield callback(request, **kwargs)
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/federation/transport/server.py", line 227, in new_func
    origin, content, request.args, *args, **kwargs
  File "/root/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/root/.synapse/local/lib/python2.7/site-packages/synapse/federation/federation_server.py", line 230, in on_context_state_request
    raise NotImplementedError("Specify an event")
exceptions.NotImplementedError: Specify an event
```

**Comment by HimanshuJanbandhu** [comment]
I would like to work on this, can someone help me please.

**Comment by hawkowl** [comment]
Fixed in #4741.
